manila/share/drivers/netapp/dataontap/client/client_cmode.py

```diff
@@ -1456,12 +1464,17 @@ def configure_active_directory(self, security_service, vserver_name):
         if security_service['ou'] is not None:
             api_args['organizational-unit'] = security_service['ou']
 
-        try:
-            LOG.debug("Trying to setup CIFS server with data: %s", api_args)
-            self.send_request('cifs-server-create', api_args)
-        except netapp_api.NaApiError as e:
-            msg = _("Failed to create CIFS server entry. %s")
-            raise exception.NetAppException(msg % e.message)
+        for attempt in range(3):
+            try:
+                LOG.debug("Trying to setup CIFS server with args: %s",
+                          api_args)
+                self.send_request('cifs-server-create', api_args)
+                return
+            except netapp_api.NaApiError as e:
+                LOG.debug("Failed to create CIFS server entry. %s", e.message)
+                time.sleep(3)
+                continue
+            raise
```

Comment:
If I read this correctly: If the 3rd attempt did not work and produced a `netapp_api.NaApiError`, it will go on without raising?
Shouldn't it raise `exception.NetAppException` like before?

Reply:
last raise on line 1477 should re-raise the last exception: https://docs.python.org/3/reference/simple_stmts.html#raise

Reply:
This `raise` is dead code.
1. Either no exception occurs and the method returned. 
2. Or the except-branch is executed and the loop continues to the next iteration until all 3 iterations are over
3. Or the error is not caught by the except-branch, because a different error occurred

```python
def configure_active_directory(func):
    for attempt in range(3):
        try:
            func()
            return
        except ZeroDivisionError:
            print("dividing by zero hurts")
            continue
        print("I will never get HERE")
        raise


def valid_print():
    print("valid statement")


def zero_division():
    1/0


def invalid_attribute():
    "a".get()


configure_active_directory(valid_print)
valid statement


configure_active_directory(zero_division)
dividing by zero hurts
dividing by zero hurts
dividing by zero hurts


configure_active_directory(invalid_attribute)
AttributeError: 'str' object has no attribute 'get'
```

I only managed to make use of the re-raise with a bulky version:

```python
def configure_active_directory_with_reraise(func):
    for attempt in range(3):
        try:
            func()
            return
        except ZeroDivisionError:
            print("dividing by zero hurts")
            if attempt == 2:
                print("I will get HERE and do the re-raise")
                raise
            else:
                continue
        print("I will never get HERE")
        raise

configure_active_directory_with_reraise(zero_division)
dividing by zero hurts
dividing by zero hurts
dividing by zero hurts
I will get HERE and do the re-raise
ZeroDivisionError: division by zero
```

I learned a lot building this simplifying examples, thanks!

Reply:
My bad, you're absolutely right! The _raise_ should be under try clause to work, like can be seen in the same module...
Thanks for finding!